esptool.py

```diff
@@ -2148,6 +2148,17 @@ def __call__(self, parser, namespace, values, option_string=None):
             except IndexError:
                 raise argparse.ArgumentError(self,'Must be pairs of an address and the binary filename to write there')
             pairs.append((address, argfile))
+
+        # Sort the addresses and check for overlapping
+        end = 0
+        for address, argfile in sorted(pairs):
+            argfile.seek(0,2)  # seek to end
+            size = argfile.tell()
```

Comment:
Because esptool erases entire sectors, the address & size can be aligned to 4096-byte boundaries when checking for overlaps - even regions which don't obviously overlap can cause data loss if they land on the same sector.

Something like:

```
address = address & ~(FLASH_SECTOR_SIZE-1)
...
end = (address + size + FLASH_SECTOR_SIZE-1) & ~(FLASH_SECTOR_SIZE-1)
```